src/dg/dg_base_state.cpp

```diff
@@ -0,0 +1,239 @@
+    this->use_auxiliary_eq = pde_physics_double->has_nonzero_diffusion;
+}
+
+template <int dim, int nstate, typename real, typename MeshType>
+real DGBaseState<dim, nstate, real, MeshType>::evaluate_CFL(std::vector<std::array<real, nstate> > soln_at_q,
+                                                            const real artificial_dissipation, const real cell_diameter,
+                                                            const unsigned int cell_degree) {
+    const unsigned int n_pts = soln_at_q.size();
+    std::vector<real> convective_eigenvalues(n_pts);
+    std::vector<real> viscosities(n_pts);
+    for (unsigned int isol = 0; isol < n_pts; ++isol) {
+        convective_eigenvalues[isol] = pde_physics_double->max_convective_eigenvalue(soln_at_q[isol]);
+        viscosities[isol] = pde_physics_double->max_viscous_eigenvalue(soln_at_q[isol]);
+    }
+    const real max_eig = *(std::max_element(convective_eigenvalues.begin(), convective_eigenvalues.end()));
+    const real max_diffusive = *(std::max_element(viscosities.begin(), viscosities.end()));
+
+    // const real cfl_convective = cell_diameter / max_eig;
+    // const real cfl_diffusive  = artificial_dissipation != 0.0 ? 0.5*cell_diameter*cell_diameter /
+    // artificial_dissipation : 1e200; real min_cfl = std::min(cfl_convective, cfl_diffusive) / (2*cell_degree + 1.0);
+
+    const unsigned int p = std::max((unsigned int)1, cell_degree);
+    const real cfl_convective = (cell_diameter / max_eig) / (2 * p + 1);  //(p * p);
+    const real cfl_diffusive = artificial_dissipation != 0.0
+                                   ? (0.5 * cell_diameter * cell_diameter / artificial_dissipation) / (p * p * p * p)
+                                   : ((this->all_parameters->ode_solver_param.ode_solver_type !=
+                                       Parameters::ODESolverParam::ODESolverEnum::implicit_solver)
+                                          ?  // if explicit use pseudotime stepping CFL
+                                          (0.5 * cell_diameter * cell_diameter / max_diffusive) / (2 * p + 1)
+                                          : 1e200);
+    real min_cfl = std::min(cfl_convective, cfl_diffusive);
+
+    if (min_cfl >= 1e190) min_cfl = cell_diameter / 1;
+
+    return min_cfl;
+}
+
+// Define a sequence of indices representing the range [1, 5]
+#define POSSIBLE_NSTATE (1)(2)(3)(4)(5)(6)
+
+// Define a macro to instantiate MyTemplate for a specific index
+#define INSTANTIATE_DISTRIBUTED(r, data, index) \
+    template class DGBaseState <PHILIP_DIM, index, double, dealii::parallel::distributed::Triangulation<PHILIP_DIM>>;
+
+#if PHILIP_DIM!=1
+BOOST_PP_SEQ_FOR_EACH(INSTANTIATE_DISTRIBUTED, _, POSSIBLE_NSTATE)
+#endif
+
+#define INSTANTIATE_SHARED(r, data, index) \
+    template class DGBaseState <PHILIP_DIM, index, double, dealii::parallel::shared::Triangulation<PHILIP_DIM>>;
+BOOST_PP_SEQ_FOR_EACH(INSTANTIATE_SHARED, _, POSSIBLE_NSTATE)
```

Comment:
very nice

Reply:
Yeah, I tried using it for some of the Physics. It's nice that it reduces the number of lines, but it can get a bit unreadable. I haven't figured out how to do 2 loops (types and nstates).

Still on the fence about whether it's better to have repetitive readable code versus preprocessor macros.